### PR TITLE
Querier: Added native multi-tenancy support

### PR DESCRIPTION
**Pull request by Abhishek357**
Signed-off-by: Abhishek357 <abhisheksinghchauhan357@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
Fixes [#3822](https://github.com/thanos-io/thanos/issues/3822) Related to [Proposal](https://github.com/thanos-io/thanos/pull/4055)
We take the tenant header and replace the tenant label if specified in the query, if not just append it in prometheus label matches using [enforceMatcher](https://github.com/prometheus-community/prom-label-proxy/blob/38b1382d7368e994e39c0cb853f3f3c9300f5045/injectproxy/enforce.go#L114)(helper function defined in prom-label proxy). Then these labels are further used to fan out a query to store API which advertises these external tenant label. Therefore achieving tenant isolation.

We use the same injection code for raw matches from the prom-label proxy, so the injection and multi-tenancy / cross tenancy are coming from the single, well-tested library.

**Alternative considered**

1. Using complete prom-label proxy as a library, importing it, and using it to modify the query itself. In Prom-label proxy we extract [expression using parser](https://github.com/prometheus-community/prom-label-proxy/blob/38b1382d7368e994e39c0cb853f3f3c9300f5045/injectproxy/routes.go#L302) from the query and then convert to matchers inside [EnforceNode](https://github.com/prometheus-community/prom-label-proxy/blob/38b1382d7368e994e39c0cb853f3f3c9300f5045/injectproxy/enforce.go#L100), then enforces label in [enforceMatcher](https://github.com/prometheus-community/prom-label-proxy/blob/38b1382d7368e994e39c0cb853f3f3c9300f5045/injectproxy/enforce.go#L114), then convert again to query. and these queries are again converted to matchers in Thanos and then further processed. Why such overhead if we are finally going to convert queries to the matcher inside Thanos and let's enforce the labels there itself using [enforceMatcher](https://github.com/prometheus-community/prom-label-proxy/blob/38b1382d7368e994e39c0cb853f3f3c9300f5045/injectproxy/enforce.go#L114).

2. Having a dedicated thanos field in [Info](https://github.com/thanos-io/thanos/blob/f4a5904b0c3be7565298cf257ad43ac1942021cc/pkg/store/proxy.go#L114) method in Store API.
    - Working smoothly by using tenant label as external label, didn't find a solid use case for this new field.

## Verification
Wrote a e2e test.

### CODE REVIEW

File: pkg/api/query/v1.go
Comment (kakkoyun) [inline pkg/api/query/v1.go:341]: It'd be nice to make this configurable.

File: pkg/query/querier.go
Comment (kakkoyun) [inline pkg/query/querier.go:270]: Could we make this configurable?

**Review by kakkoyun — CHANGES_REQUESTED** [review]
Looking good so far. 
It'd be amazing to create a wrapper around existing querier, such as QuerierWithTenant. It'd make the code cleaner and more maintainable.

File: pkg/api/query/v1.go
Comment (bwplotka) [inline pkg/api/query/v1.go:341]: `X-` prefix also?

File: pkg/query/querier.go
Comment (bwplotka) [inline pkg/query/querier.go:270]: Here, we could immidately filter out if someone is asking for another tenant data?

Comment (bwplotka) [reply pkg/query/querier.go:270]: +1

File: pkg/query/querier.go
Comment (bwplotka) [inline pkg/query/querier.go:269]: This is security impacting code, so it would be best to put it into a helper or even import a production-grade promllabel proxy library for this work (and test properly).

**Review by bwplotka — COMMENTED** [review]
High-level questions (maybe even proposal)

* Injecting identifier what tenant is making the call (to any API)
* Injecting identifier of data that we have to scope to (single-tenant, multiple in future) - I guess this is the purpose of this PR?

* Headers have to be configured, the same tenant label (avoid hardcoded strings)
* Can we reuse promlabel. It's used mainly for PromQL mangling, however, we could decouple injection and have the same injection code for PromQL and raw matches, so the injection and multi-tenancy / cross tenancy are coming from the single, well-tested library. That we can use in Thanos, Observatorium, PromLabel proxy sidecar it. One source of truth.

File: pkg/api/query/v1.go
Comment (Abhishek357) [reply pkg/api/query/v1.go]: Yes, let's make the prefix as well configurable to be consistent with the receiver.

**Comment by Abhishek357** [comment]
Blocked by [#66](https://github.com/prometheus-community/prom-label-proxy/pull/66)

File: cmd/thanos/query.go
Comment (yashrsharma44) [inline cmd/thanos/query.go:156]:
```suggestion
	tenantHeader := cmd.Flag("query.tenant-header", "HTTP header to determine tenant access for a query.").Default(DefaultTenantHeader).String()
```

File: cmd/thanos/query.go
Comment (yashrsharma44) [inline cmd/thanos/query.go:158]:
```suggestion
	tenantLabelName := cmd.Flag("query.tenant-label-name", "Tenant label name in query.").Default(DefaultTenantLabel).String()
```

File: pkg/promclient/promclient.go
Comment (yashrsharma44) [inline pkg/promclient/promclient.go:124]: Are we using any default tenant header, in case `tenantHeader` is empty string?

File: pkg/promclient/promclient.go
Comment (yashrsharma44) [inline pkg/promclient/promclient.go:169]:
```suggestion
	body, _, err := c.req2xx(ctx, &u, http.MethodGet, EMPTY_HEADER)
```
You can define a constant for empty tenant header, would be easy to read and understand what does `""` means.
Feel free to name variable other than `EMPTY_HEADER` :)

File: pkg/query/querier.go
Comment (yashrsharma44) [inline pkg/query/querier.go:15]: `ef` sounds too abstract, can you leave the import as it is?

File: pkg/query/querier.go
Comment (yashrsharma44) [inline pkg/query/querier.go:287]: stray print lines? Can be removed, if of no use :)

File: test/e2e/query_test.go
Comment (yashrsharma44) [inline test/e2e/query_test.go:530]: can you add in a constant for empty string? Hard to read and understand what is `""` :)

File: test/e2e/query_test.go
Comment (yashrsharma44) [inline test/e2e/query_test.go:623]: We need to add in this also for checking if creation of receiver was successful.
```suggestion
	receiver1, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "1", 1)
	testutil.Ok(t, err)
```

File: test/e2e/query_test.go
Comment (yashrsharma44) [inline test/e2e/query_test.go:624]:
```suggestion
	receiver2, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "2", 1)
	testutil.Ok(t, err)
```

Comment (yashrsharma44) [reply test/e2e/query_test.go:624]: Seems like muse bot is also complaining about the same thing 😂

File: test/e2e/query_test.go
Comment (yashrsharma44) [inline test/e2e/query_test.go:645]: Did this got answered?
Described here - https://github.com/thanos-io/thanos/blob/main/pkg/query/storeset.go#L151

**Review by yashrsharma44 — CHANGES_REQUESTED** [review]
Looks good, I need to test it out to cross-check the findings :)

**Comment by Abhishek357** [comment]
> Looks good, I need to test it out to cross-check the findings :)

We can not test it until we merge PR in prom-label-proxy because only then we will be able to use those enforcing helper functions.

File: pkg/promclient/promclient.go
Comment (Abhishek357) [reply pkg/promclient/promclient.go:124]: No, we are not using any default tenant header value.
In case tenantHeader is an empty string we assume that the user has access to tenant it specifies.

File: pkg/query/querier.go
Comment (Abhishek357) [reply pkg/query/querier.go:15]: agree that ef is too abstract, but let's relabel it else it will be used as injectproxy and IMO it's kind of misleading because it is no longer used as a proxy. WDYT :)

File: cmd/thanos/query.go
Comment (kakkoyun) [inline cmd/thanos/query.go:55]: nit: Maybe we should move these to a more common package since also the receiver uses these constants.

File: cmd/thanos/query.go
Comment (kakkoyun) [inline cmd/thanos/query.go:156]: nit: Shared behaviour can be moved to a common package.

File: pkg/query/querier.go
Comment (kakkoyun) [inline pkg/query/querier.go:36]: Maybe we can create a struct for the tenant?

File: cmd/thanos/query.go
Comment (kakkoyun) [reply cmd/thanos/query.go:156]: maybe we can  drop the `query` prefix

File: pkg/query/querier.go
Comment (kakkoyun) [inline pkg/query/querier.go:271]: We can move to the constructor.

File: pkg/api/query/v1.go
Comment (kakkoyun) [inline pkg/api/query/v1.go:349]: This could be unlined, no?

File: pkg/api/query/v1.go
Comment (kakkoyun) [inline pkg/api/query/v1.go:470]: ditto, and many other usages of the exactly same statement.

File: pkg/api/query/v1.go
Comment (kakkoyun) [inline pkg/api/query/v1.go:129]: Why don't we initiate `TenantInfo` where we call `NewQueryAPI`?

File: pkg/promclient/promclient.go
Comment (kakkoyun) [inline pkg/promclient/promclient.go:168]: Why don't we just pass nil?

File: pkg/promclient/promclient.go
Comment (kakkoyun) [inline pkg/promclient/promclient.go:123]: You should check if the `tenantHeader` is in the map or not.

File: pkg/promclient/promclient.go
Comment (kakkoyun) [inline pkg/promclient/promclient.go:512]: We should be able to pass nil/empty map here.

File: pkg/promclient/promclient.go
Comment (kakkoyun) [inline pkg/promclient/promclient.go:594]: ditto

File: pkg/query/querier.go
Comment (kakkoyun) [inline pkg/query/querier.go:41]:
```suggestion
type QueryableCreator func(deduplicate bool, replicaLabels []string, storeDebugMatchers [][]*labels.Matcher, maxResolutionMillis int64, partialResponse, skipChunks bool, tenant *TenantInfo) storage.Queryable
```
Why not use a pointer? So that we can just pass nil to indicate tenant doesn't exist.

File: pkg/query/querier_test.go
Comment (kakkoyun) [inline pkg/query/querier_test.go:46]:
```suggestion
	queryable := queryableCreator(false, nil, nil, oneHourMillis, false, false, nil)
```
See the above comment.

**Review by kakkoyun — CHANGES_REQUESTED** [review]

**Comment by kakkoyun** [comment]
Hey @Abhishek357 could you please address the review comments and resolve them afterwards? Also it'd be nice to rebase this PR.

File: pkg/api/query/v1.go
Comment (Abhishek357) [reply pkg/api/query/v1.go:129]: Because we want to propagate only valuable TenantInfo, we get this when an HTTP request comes and we extract the tenant access according to the configurable tenant Header name (it's not necessary to propagate this label name).

**Comment by Abhishek357** [comment]
@kakkoyun friendly ping addressed all the comments and this PR is good to go. Test failing is not related.

File: pkg/api/query/v1.go
Comment (yashrsharma44) [inline pkg/api/query/v1.go:696]:
```suggestion
	q, err := qapi.queryableCreate(true, nil, storeDebugMatchers, 0, enablePartialResponse, true, &query.TenantInfo{
		Access:    r.Header.Get(qapi.tenantHeader),
		LabelName: qapi.tenantLabelName,
	}).
```

File: pkg/promclient/promclient.go
Comment (yashrsharma44) [inline pkg/promclient/promclient.go:125]:
```suggestion
        for _, value := range(headers){
            req.Header.Set("THANOS-TENANT", value)
        }
```

**Review by yashrsharma44 — COMMENTED** [review]
Some suggestions

File: pkg/promclient/promclient.go
Comment (yashrsharma44) [inline pkg/promclient/promclient.go:594]: Use a pointer in `req2xx` and pass nil
```suggestion
	body, _, err := c.req2xx(ctx, &u, http.MethodGet, nil)
```

**Review by yashrsharma44 — COMMENTED** [review]
Some more changes

File: pkg/promclient/promclient.go
Comment (Abhishek357) [reply pkg/promclient/promclient.go]: got the idea!

File: go.mod
Comment (Abhishek357) [reply go.mod:66]: @Lift-dev ignore

Comment (Abhishek357) [reply go.mod:66]: @Lift-dev ignore

File: go.mod
Comment (bwplotka) [reply go.mod:54]: @sonatype-lift help

Comment (bwplotka) [reply go.mod:54]: help

**Comment by bwplotka** [comment]
This sonar type lift thingy is misleading. Those transient versions are not used in most cases, so there is no action required here.

**Comment by bwplotka** [comment]
Can you look how we can configure this to ignore those kind of false alarms? @Abhishek357 ?

File: go.mod
Comment (TomMD) [reply go.mod:54]: hello @bwplotka, am I right in saying this error is not sufficiently detailed?

**Comment by TomMD** [comment]
> Can you look how we can configure this to ignore those kind of false alarms? @Abhishek357 ?

@Abhishek357 and @bwplotka you can add a `.lift/config.toml` file with a field of:

```
ignoreRules = [ "Severe OSS Vulnerability", "Critical OSS Vulnerability" ]
```

Which should screen out these results.  You can also look at the full [configuration reference](https://help.sonatype.com/lift/configuration-reference) for more options.

**Comment by ChanderG** [comment]
@bwplotka @Abhishek357 Looks like this should be merged in? Definitely a useful feature and the conversation above seems to indicate no problems/blockers.

**Comment by shaardie** [comment]
I think this is still relevant. I am also a bit confused this is closed and not merged?!

**Pull request closed without merging**
